### ISSUE TITLE
deepseek_v41: the indexer's top-k depends on whether a position was prefilled or resumed

### ISSUE THREAD

**Issue by JustVugg** [JustVugg/colibri#1554]
Found while wiring KV prefix reuse into this engine (#1553 does qwen36 and olmoe; V4.1 is left out because of this).

## What happens

Feed a sequence as one prefill, or as a prefill of the first `k` tokens followed by a second forward that resumes at position `k`. Every piece of per-position state resumes exactly: the window ring and its position map, the compressed KV and index keys, the compressor's partial group, the engram history. The generated tokens still diverge, a few tokens in.

The cause is the indexer. With `index_topk` raised until nothing is selected away, the two feedings agree exactly at every split I tried. With the fixture's `index_topk=4` they do not.

```
tiny fixture (tools/make_dsv41_tiny.py), V41_DSPARK=0, prompt of ~100 byte-tokens

index_topk = 4                       index_topk = 128
prefix  96, tail   4: DIFF           prefix  96, tail   4: SAME
prefix  88, tail  12: DIFF           prefix  88, tail  12: SAME
prefix  76, tail  24: DIFF           prefix  76, tail  24: SAME
prefix  52, tail  48: SAME           prefix  52, tail  48: SAME
prefix  50, tail  50: SAME           prefix  50, tail  50: SAME
```

So for the same position, the set of compressed groups it attends to is not the same when that position is computed inside a prefill chunk as when it is computed in a chunk that resumes mid-sequence. `attention_run` says as much in passing: during prefill the window half of `kv_all` is the whole chunk, while a resumed chunk gets the `window` ring, and `indexer_run` is handed that count.

## Why it is invisible today

A turn is always one prefill at position 0 followed by incremental steps, so the two regimes never describe the same position. Prefix reuse is what makes them overlap. It is still a question about the engine rather than about reuse: one of the two paths does not agree with the other about what a query may see, and the tiny oracle validates each regime separately without ever comparing them.

## Reproduction

Two turns over the serve protocol on the tiny fixture, the second extending the first, comparing against a cold engine given the extended prompt directly. I have the harness on a local branch together with three other fixes this turned up, which are worth keeping whatever is decided here:

- `ring_save` and `cstate_save_*` are sized for one speculative draft block. They are written whenever a forward has more than one row and starts past position 0, which until now only ever meant the verify batch. A resumed prefill is many rows past position 0 and overflows both (heap-buffer-overflow, caught by ASan).
- `spec_step` decides "seed the draft stages, do not draft" by `start_pos == 0`, and the seeding call passes 0 as the position. A resumed prefill therefore falls through into the drafting path with a NULL draft buffer (null deref), and would seed the stages at the wrong positions.

## What I would do next

Decide which path is right against the vendor reference, make the other match, and add a gate that feeds one sequence both ways and compares. Then V4.1 joins #1553's contract unchanged.

**Comment by JustVugg** [comment]
Correction to the measurement in the report above, and the root cause.

The numbers I posted were taken against an engine that had no reuse wired in at all (a branch switch had reverted the file under me), so "identical" was guaranteed and told us nothing. Re-measured properly, here is what is actually happening.

**The cause.** The index-key schedule is built for `n > 1 && start_pos > 0`. That condition means "the speculative verify batch", and it only worked because a prefill always started at position 0. A prefill that resumes mid-sequence matches it too and gets decode semantics: its rows read another layer's published keys instead of their own index owner's, so the same position selects a different index top-k than it selected when prefilled cold. Guarding on `all_logits` instead (spec_verify is its only caller), and on `n == 1` for the single-position branch, makes a resumed prefill reproduce a cold one exactly at every split I tried (tails of 4, 12, 24, 48 and 50 tokens), with the oracle still 8/8 at caps 1, 8 and 49.

**What remains, and why it is not a bug.** Both rules are the vendor's and neither can be moved:

| | tiny oracle |
|---|---|
| as shipped | 8/8 |
| prefill given the decode schedule | 7/8 |
| decode given the prefill owner | 1/8 |

So a position is attended differently depending on whether it was prefilled or decoded. A reused prefix in a chat contains the previous turn's generated tokens, so it cannot reproduce a cold prefill of the same text. Five of the six scenarios in the shared contract pass on this engine; the one that replays the reply does not, and that is the chat shape.

Reuse here is therefore a behaviour decision rather than a wiring job: the resumed state is the true sequential one, but the same question asked in a fresh session and as a second turn can differ. Holding it until that is decided.

The three fixes this turned up stand on their own and are on the work branch: the two undo buffers sized for a draft block, the `spec_step` seeding guard, and the record following the speculative rollback.

**Comment by JustVugg** [comment]
Fixed in #1558 (merged into dev as b5968eb6).

The root cause was the index-key schedule being selected by `n > 1 && start_pos > 0`, a condition that meant "the speculative verify batch" only because a prefill always started at position 0. It is now selected by `all_logits`, and a resumed prefill reproduces a cold one exactly at every split point, with the oracle unchanged at 8/8. Three more fixes came with it: the two undo buffers sized for a draft block, the `spec_step` seeding guard, and the record following the speculative rollback.

What does not go away is the part that belongs to the model: a position attends differently depending on whether it was prefilled or decoded, and both rules are load-bearing against the reference. So reuse on this engine is opt-in (`COLI_KV_PREFIX=1`), with the reason and the measurements in docs/deepseek-v41.md.
